### DISCUSSION TITLE
Summary First Protocol

### DISCUSSION THREAD

**Discussion by manogyasingh** [cline/cline#671] [Feature Requests]
I suggest following a convention that alongside a code file, a summary file also be made (app.py > app.py.summary). This will help agents (human or LLM) to get context on how to integrate different modules/files etc without filling up the context window. How it would work is:

### SUMMARY FIRST PROTOCOL
1. Instruct the agent to always create a file.summary whenever creating a file and put in it a very short overview of all the exposed functions (only the name, purpose,  input and output formats, and any other VERY important detail)
2. Instruct the agent to always check for and read file.summary instead of file to understand the functions/methods etc in the file.
3. Instruct the agent to always update file.summary right after updating any file.
4. Section summaries and a project summary must also be maintained similarly. These will be periodically updated and will only contain the methods we intend to expose from that section.
5. All summary files will be tracked by version control.
6. For existing codebases, implement a functionality to be able to generate such summaries for the entire codebase with a single command with AI.

This aims to solve the problem of modularisation. Humans are good at working with large codebases because we can efficiently keep only the requisite knowledge in working memory at any given point. For example someone working on frontend only needs to know the APIs exposed by the backend, not its entire structure. But currently LLMs needlessly exhaust their context window digesting whole codefiles just to see how one function in it works. SFP fixes this by giving agents a tiny documentation to refer to instead of the whole code.

Another important challenge this addresses is that as codebases scale, compiling a comprehensive documentation gets harder and harder, by summarising on the fly we avert this scenario. Making a human friendly documentation around this short summary would be trivial.

Note that this is different from RAG because RAG too has inefficiencies as its indexing and searching the whole code, when all one needs is two lines per function. The summary will have everything needed and nothing not needed to use the things in that file.

**Comment by tyrale** [comment]
I stumbled into something similar this week. I have it maintain and update a tasks.md file. it holds crucial working knowledge, structure, tasks, and bugs. It's been the break through I've been looking for when working on larger projects. I love the summary idea. I'll going to add that in my process. 

I do believe this has been so effective that it needs to become a feature of the Agent. At the root it's about smaller files that help track the project and deliver information to the AI to avoid trying to consume or hold the full project in "memory".

**Comment by tyrale** [comment]
I created a py to handle the summary instrucitons and added a git hook to trigger it to run before each commit.
